switchy/connection/aioesl.py

```diff
@@ -0,0 +1,237 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# Copyright (c) 2017 Tyler Goodlet <tgoodlet@gmail.com>
+"""
+Asyncio ESL connection abstactions
+"""
+import asyncio
+import time
+from functools import partial
+from threading import get_ident
+from . import ConnectionError
+from .. import utils
+from ..protocol import InboundProtocol
+
+
+async def await_in_order(awaitables, loop, timeout=None):
+    awaitables = map(partial(asyncio.ensure_future, loop=loop), awaitables)
+    for awaitable in awaitables:
+        try:
+            res = await asyncio.wait_for(awaitable, timeout=timeout, loop=loop)
+        except (asyncio.CancelledError, asyncio.TimeoutError) as err:
+            for awaitable in awaitables:
+                awaitable.cancel()
+            raise
+
+    return res
```

Comment:
This now looks like you've also duplicated something very close to what's in the standard library: `asyncio.wait`, run with `FIRST_COMPLETED` that cancels all remaining if the first result is a `CancelledError`:

```python
pending = map(partial(asyncio.ensure_future, loop=loop), awaitables)
while pending:
    try:
        done, pending = yield from asyncio.wait(pending, return_when=FIRST_COMPLETED)
    except (asyncio.CancelledError, asyncio.TimeoutError) as err:
        for awaitable in pending:
            pending.cancel()
        raise
```

Reply:
> Wait for the Futures and coroutine objects given by the sequence futures to complete. Coroutines will be wrapped in Tasks. Returns two sets of Future: (done, pending).

Well, technically it doesn't wait for them in order, and the `map(partial(asyncio.ensure_future, ...), ...)` is unnecessary as it does it internally.

But this looks like you've started all the tasks at basically the time and then await them in order. But if order doesn't matter, then `wait` is better because you'll be more optimistic with your cancelling.

Otherwise I think the ensure_future needs to move into the loop.

Reply:
`asyncio.wait` [doesn't await futures in order](https://github.com/python/cpython/blob/master/Lib/asyncio/tasks.py#L307) due to using a `set()` for iteration.

Reply:
Oh, I've forgotten that `map` is lazy, so yeah, maybe its not the same thing.